### PR TITLE
Minimal changes to support constexpr allocation in MSVC

### PR DESCRIPTION
**Pull request by CaseyCarter**
* Define `_CONSTEXPR20_DYNALLOC` macro to `constexpr` when the compiler defines `__cpp_constexpr_dynamic_alloc` and `inline` otherwise.
* Implement and test `ranges::construct_at` and `ranges::destroy_at`, mark them (and `std` flavors) `_CONSTEXPR20_DYNALLOC`.
* Implement `ranges::destroy` and `ranges::destroy_n` which share machinery with `ranges::destroy_at` with minimal test coverage. (More to follow.)

[This is a dual of internal MSVC-PR-275909.]

Partially implements #37 and #39.

### CODE REVIEW

**Comment by CaseyCarter** [comment]
Full disclosure: there's really no reason for `destroy_n` to be here other than that it was sitting on the same branch with the rest of this, waiting for test coverage that still isn't here. I could pull it, but I'd prefer to leave this altogether and follow it with a test coverage PR.

File: stl/inc/memory
Comment (StephanTLavavej) [inline stl/inc/memory:397]: There are 3 remaining occurrences of `/* _CONSTEXPR20_DYNALLOC */`; is this intentional?

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/memory
Comment (CaseyCarter) [reply stl/inc/memory:397]: There are 3 in `<memory>` and 4 in `<xmemory>`. They mark functions that don't need to be `constexpr` to support `construct_at` & `destroy_at`. The 3 in `<memory>` will be uncommented when I checkin comprehensive constexpr test coverage for `ranges::destroy`/`ranges::destroy_n`. The 4 in `<xmemory>` are there from my trial `constexpr` implementation of `ranges::uninitialized_move`. I left them in place since I assumed we'd want to make those things `constexpr` to support containers, but that's really just silly: we can mark them later when needed.

Chopping the 4 and leaving the three.

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:291]: HM in the uninitialized_meow PR we introduced a voidify Funktion. If it is not too far out we should wait until it is merged or reproduce it here

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:352]: I believe that is not necessary as there is no requires clause

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:310]: Putting this constraint in the template header instead of a trailing requires clause seems more conventional. Was there a specific reason here?

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:460]: Could we separate the decrement as we do in the other meow_n functions

I am totally fine with changing to this style but consistency

**Review by miscco — COMMENTED** [review]
Blah blah missing test coverage, but this looks great

File: stl/inc/memory
Comment (CaseyCarter) [reply stl/inc/memory:291]: The whole point of pushing this through before finishing test coverage is to avoid blocking the compiler team. If we get `uninit_meow` merged in time, I'll make the `voidify` change here. Otherwise, I'll do so in the followup. (Leaving this comment open as a reminder.)

File: stl/inc/memory
Comment (CaseyCarter) [reply stl/inc/memory:352]: Good catch - I'll audit all the clang-format annotations.

File: stl/inc/memory
Comment (CaseyCarter) [reply stl/inc/memory:460]: I should probably `do`...`while` this like `std::destroy_n` since we know `_Count` is greater than `0` here. Actually I should steal the `is_trivially_destructible` optimization while I'm at it.

File: stl/inc/memory
Comment (CaseyCarter) [reply stl/inc/memory:310]: > Was there a specific reason here?

Yes: using the function parameters to be lazy and avoid writing a parameter-list for the requires-expression. (I didn't say it was a _good_ reason.)

I'll clean this up.

File: stl/inc/memory
Comment (miscco) [reply stl/inc/memory:291]: :+1:

File: stl/inc/memory
Comment (miscco) [reply stl/inc/memory]: I love me suggesting something tiny and inconsequential and you making things 10 times better in the process

File: stl/inc/memory
Comment (miscco) [inline stl/inc/memory:460]: No change requested. I hate do while loops. 

It is even 2 lines longer. Think about all that screen real estate

**Comment by miscco** [comment]
Does that mean we can have a look at constexpr container after this is merged (excluding string and vector of course)

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/memory
Comment (StephanTLavavej) [reply stl/inc/memory:460]: Indeed, `do-while` loops aren't our preferred form - but we'll use them when they can potentially save cycles. They're a signal that we've already established the loop invariant at the beginning.

File: stl/inc/memory
Comment (StephanTLavavej) [reply stl/inc/memory:291]: MSVC-PR-276291 is internally out for review, containing `uninit_meow`, so it's a race :rocket:

File: stl/inc/memory
Comment (CaseyCarter) [reply stl/inc/memory:460]:
```c++
for (...) {
}
```
2 lines.
```c++
while (...) {
}
```
2 lines.
```c++
do {
} while(...);
```
2 lines.

Fun aside, like Stephan says we use do/while when the invariant has already been validated elsewhere like this,

**Review by mnatsuhara — APPROVED** [review]

File: stl/inc/memory
Comment (CaseyCarter) [reply stl/inc/memory:291]: I've annotated the `destroy`/`destroy_n` test coverage followup work item in #39 to mention integration with GH-1164 and `_Voidify_iter` specifically.

**Pull request merged**
